September 2025 new azimuthal average changes

Heidy-Elkhaligy opened the pull request:
image.cpp added AverageRotationally and ApplyRampFilter( ) functions azimuthal_average.cpp  now contains diameTR,  RASTR and SPOT-RASTR  options

# Description
Old version: Alignment was performed via iterative real-space cross-correlation with angle scanning until convergence.
New version: Alignment now supports an initial alignment step using autocorrelation and Fourier Transform (FT), which generates a reference image for a second round of alignment using cross-correlation, providing more robust and faster convergence
New additions in the new version:
1. Diameter determination (diameTR): Automatic detection of tube boundaries and support for binning by tube diameter.
2. Added RASTR (Reconstruction of Average Subtracted Tubular Regions) option, which isolates small areas of tubular architecture by subtracting a masked azimuthal average.
3. Added SPOT-RASTR (Specimen Preparation On a Tube – RASTR) to subtract the signal of the lipid nano-tubes


Fixes # (issue)

# I have rebased my feature branch to be current with the master branch using to minimize conflicts and headaches

- [x] yes
- [ ] no

# Which compilers were tested

- [ ] g++
- [x] icpc
- [ ] clang
- [ ] other (please specify)

# These changes are isolated to the

- [ ] gui
- [x] core library
- [ ] gpu core library
- [x] program it modifies

# How has the functionality been tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Tested manually from GUI
- [x] Tested manually from CLI
- [ ] Passed console tests
- [ ] Passed samples functional testing
- [ ] other (please specify)

# Checklist:

- [x] I have not changed anything that did not need to be changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, (***w.r.t. why***), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/bHimes/cisTEM_docs) {Ok to pass for now}
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

Heidy-Elkhaligy committed:
September 2025 new azimuthal average changes
image.cpp added AverageRotationally and ApplyRampFilter( ) functions
azimuthal_average.cpp  now contains RASTR and SPOT-RASTR  options

bHimes reviewed (changes requested):
I assume these methods do what they should. My comments are mostly to clean up some easy performance issues with memory allocations (calls to new) and narrowing the scope of local variables which is safer and allows the compiler to better optimize register usage.

bHimes commented:
Hi @Heidy-Elkhaligy -

I saw a force pushed commit and several conversations marked resolved that were not resolved. I hope this is just an issue with using git/github (which can be a pain in the butt at first.)

If you have any questions about the code review process or the mechanics of how to respond to/push up changes to code review requests, please let me know. I'm happy to also set aside 10-15 min for a quick Zoom to discuss.

If on the other hand, you did mean to mark my comments as resolved without addressing the code or addressing the comments either, please let me know. If you aren't planning on future contributions and just want to "get this done" I'm happy to just accept your PR and then fix the code myself, which is much easier than going through code review.

twagner9 commented:
> Hi @Heidy-Elkhaligy -
> 
> I saw a force pushed commit and several conversations marked resolved that were not resolved. I hope this is just an issue with using git/github (which can be a pain in the butt at first.)
> 
> If you have any questions about the code review process or the mechanics of how to respond to/push up changes to code review requests, please let me know. I'm happy to also set aside 10-15 min for a quick Zoom to discuss.
> 
> If on the other hand, you did mean to mark my comments as resolved without addressing the code or addressing the comments either, please let me know. If you aren't planning on future contributions and just want to "get this done" I'm happy to just accept your PR and then fix the code myself, which is much easier than going through code review.
> 
> I only made comments on changes to the Image class as this might affect the project more broadly. If you would like a review of src/programs/azimuthal_average/azimuthal_average.cpp I can do that as well, but am perfectly happy to let that exist in whatever state you find acceptable as it only affects itself.

Sorry @bHimes, I am working with @Heidy-Elkhaligy on this PR in-office -- we have been going through the comments you've suggested and implementing them, @Heidy-Elkhaligy is also just going through some of her azimuthal_average.cpp code as well to check if there's anything else she wants to correct there before pushing.

Apologies for not letting you know sooner, and the changes will probably be pushed sometime this week!

Heidy-Elkhaligy committed:
cleaning up

1. Renamed variables for clarity:
   - switchBackToComplex → input_image_in_fourier_space
   - do_reverse_FFT → do_backward_FFT
2. Replaced raw pointers with smart pointers and removed delete[] calls
3. Restricted loop variables to local scope
4. Optimized powf/sqrtf usage by removing redundant powf calls

Heidy-Elkhaligy committed:
Additional cleaning
1. Renamed variables for clarity
2. Replaced some raw pointers with smart pointers and removed delete[] calls
3. Restricted loop variables to local scope

Heidy-Elkhaligy commented:
@bHimes, I implemented some of the changes you suggested in image.cpp within azimuthal_average.cpp. Sorry for the force push; it was necessary because we had to rebase. I’d appreciate your help reviewing those updates.

Heidy-Elkhaligy commented:
@bHimes I'm also going on a month-long vacation starting this week. Therefore, there's no rush to review the code, as I won't be able to make any changes during that time. Thank you again for your help and support.

bHimes commented:
Awesome! My kinda traveler. Happy to wait if you'd like to learn more about the process. If you like, i'm also happy to pull in your changes and make any adjustments for you

Heidy-Elkhaligy commented:
Awesome, thank you! I’d love to learn more about the process and the logic behind the changes so I can avoid making the same mistakes in the future. I’m not a programmer or computer scientist, and my coding skills are pretty basic, so understanding the reasoning would really help me improve. If you decide to go ahead and finish the pull request, I’d really appreciate it if you could point out what I missed or did incorrectly, that would be super helpful for me to learn from!
Thank you for your patience and support!

timothygrant80 commented:
@bHimes - are you happy with the state of this now?

bHimes commented:
i'll look it over again tomorrow

bHimes commented:
Included in PR #572, only modified RampFilter method in the Image class, otherwise, the other changes @Heidy-Elkhaligy made looked fine.
